### PR TITLE
Limit all records extracted to before 2021

### PR DESCRIPTION
**Pull request by ThePletch**
## Why

[Relevant ticket](https://broadinstitute.atlassian.net/browse/DSPDC-1440)

## This PR
* Adds a filter to all three pipelines ensuring that no records submitted after 2020 are included.
* Replaces the largely unused dateRangeBegin and dateRangeEnd filters with ones that work off field filtering.

### CODE REVIEW

File: hles/extraction/src/main/scala/org/broadinstitute/monster/dap/CslbExtractionPipeline.scala [hunk 1]
```diff
@@ -15,9 +15,18 @@ object CslbExtractionPipeline extends ScioApp[Args] {
...
+      FilterDirective("canine_social_and_learned_behavior_complete", FilterOps.==, "2")
+    ) ++
+      args.startTime
+        .map(start =>
+          List(FilterDirective("cslb_date", FilterOps.>, RedCapClient.redcapFormatDate(start)))
+        )
+        .getOrElse(List()) ++
+      args.endTime
+        .map(end => List(FilterDirective("cslb_date", FilterOps.<, RedCapClient.redcapFormatDate(end))))
+        .getOrElse(List())
```
Comment (ThePletch) [inline hles/extraction/src/main/scala/org/broadinstitute/monster/dap/CslbExtractionPipeline.scala:29]: i am positive that there's a simpler way to conditionally add a value to a list inline, but i don't know enough scala to know it

File: hles/extraction/src/main/scala/org/broadinstitute/monster/dap/FilterDirective.scala [hunk 2]
```diff
@@ -23,4 +23,14 @@ object FilterOps {
...
   * Their API expects directives to be of the form [field][op][value]
   */
-case class FilterDirective(field: String, operation: FilterOp, comparand: String)
+case class FilterDirective(field: String, operation: FilterOp, comparand: String) {
+
+  // wrap any comparands with a space in quotes
+  def spaceEscapedComparand: String = {
+    comparand match {
+      case comparandWithSpace: String if comparandWithSpace.contains(" ") =>
+        s""""${comparandWithSpace}""""
```
Comment (ThePletch) [inline hles/extraction/src/main/scala/org/broadinstitute/monster/dap/FilterDirective.scala:32]: `s"\"${comparandWithSpace}\""` runs into an eight-year standing (!!!) bug with scala where character escaping is completely broken in s-strings.

File: hles/extraction/src/main/scala/org/broadinstitute/monster/dap/RedCapClient.scala [hunk 3]
```diff
@@ -35,22 +37,20 @@ object RedCapClient {
   /** URL for the production RedCap API. */
   private val apiRoute = "https://redcap.dogagingproject.org/api/"
 
-  /** Formatter matching the production RedCap's interface. */
-  private val dateFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
```
Comment (ThePletch) [inline hles/extraction/src/main/scala/org/broadinstitute/monster/dap/RedCapClient.scala:39]: what fools we  were

**Review by ThePletch — COMMENTED** [review]
self cr, requests for comment in a few places

File: hles/extraction/src/main/scala/org/broadinstitute/monster/dap/FilterDirective.scala (hunk 2 above)
Comment (aherbst-broad) [reply hles/extraction/src/main/scala/org/broadinstitute/monster/dap/FilterDirective.scala:32]: Great find, can you add any color to this?

File: hles/extraction/src/main/scala/org/broadinstitute/monster/dap/HLESurveyExtractionPipeline.scala [hunk 4]
```diff
@@ -8,7 +8,7 @@ import Args._
 /** Entry-point for the HLE extraction pipeline. */
 object HLESurveyExtractionPipeline extends ScioApp[Args] {
 
-  val HLESEpoch = "2018-01-01"
+  val HLESEpoch = "2018-01-01 00:00"
```
Comment (aherbst-broad) [inline hles/extraction/src/main/scala/org/broadinstitute/monster/dap/HLESurveyExtractionPipeline.scala:11]: ⛏️ For consistency, we should make this an OffsetDateTime and pass it through the `redCapFormatDate` method like the CLI args

File: hles/extraction/src/main/scala/org/broadinstitute/monster/dap/CslbExtractionPipeline.scala (hunk 1 above)
Comment (aherbst-broad) [reply hles/extraction/src/main/scala/org/broadinstitute/monster/dap/CslbExtractionPipeline.scala:29]: I am not sure myself; a case match might be slightly more idiomatic?

```suggestion
  List(
      FilterDirective("canine_social_and_learned_behavior_complete", FilterOps.==, "2")
    ) ++
      (args.startTime match {
        case Some(start) =>
          List(FilterDirective("cslb_date", FilterOps.>, RedCapClient.redcapFormatDate(start)))
        case _ => List()
      }) ++
      (args.endTime match {
        case Some(end) =>
          List(FilterDirective("cslb_date", FilterOps.<, RedCapClient.redcapFormatDate(end)))
        case _ => List()
      })
```

**Review by aherbst-broad — APPROVED** [review]
Good finds

**Review by quazi-broad — APPROVED** [review]
Was not expecting this issue to be so extensive. Good work!

File: hles/extraction/src/main/scala/org/broadinstitute/monster/dap/CslbExtractionPipeline.scala (hunk 1 above)
Comment (ThePletch) [reply hles/extraction/src/main/scala/org/broadinstitute/monster/dap/CslbExtractionPipeline.scala:29]: if memory serves, scala best practices discourage using `match` on `Option` types in favor of the `map/getOrElse` pattern i'm using here

File: hles/extraction/src/main/scala/org/broadinstitute/monster/dap/FilterDirective.scala (hunk 2 above)
Comment (ThePletch) [reply hles/extraction/src/main/scala/org/broadinstitute/monster/dap/FilterDirective.scala:32]: https://github.com/scala/bug/issues/6476 bug info is here, dating back to october 2012

**Pull request merged**